### ISSUE TITLE
TS can implicitly resolve some variables to `any` in impossible situations whether `noImplicitAny` is set or not

### ISSUE THREAD

**Issue by EdwardDrapkin** [microsoft/TypeScript#37026]
**TypeScript Version:**  3.8-Beta

**Search Terms:**
any
conditional types

**Expected behavior:**
If you type a generic parameter that's impossible for TS to resolve because it doesn't have a totality checker, TS will type that variable as 'any'.  For example, if you define a generic factory function interface and try to accept parameters that are returned by said interface, TS appears to get confused, type as `any` and bail, whether `noImplicitAny` is enabled or not. 

**Actual behavior:**
Variables are typed as 'any' when `noImplicitAny` is disabled and an error is raised when it isn't.  Ideally, even in the `noImplicitAny` case, a warning would be raised because that sort of type is most likely a code smell.

**Code**
```ts
interface FactoryFunction<T> {
  (): T
}

function onlyAcceptsFactoryProducts<T extends FactoryFunction<any>>(foo: T extends FactoryFunction<infer U> ? U : never): T extends FactoryFunction<infer U> ? U : never {
  return foo;
}

// both are typed as any
const a = onlyAcceptsFactoryProducts('foo');
const b = onlyAcceptsFactoryProducts(123);
```

<details><summary><b>Output</b></summary>

```ts
"use strict";

function onlyAcceptsFactoryProducts(foo) {

    return foo;

}

// both are typed as any

const a = onlyAcceptsFactoryProducts('foo');

const b = onlyAcceptsFactoryProducts(123);


```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](http://www.typescriptlang.org/play/?ts=3.8-Beta#code/JYOwLgpgTgZghgYwgAgGKLAeygT1QVxATGExAB4AVAPmQG8AoZZACgEoAuZShgXwYYxCxUiGRkANjgCCCJAAcwAZ3TFsOAApRMAE3zElVZBAAekEDqVoM6gkRJlycEDmrUWMTJi6VjZiBZWqli4diKOoDDQyACqtAD8schcIBAAbtCc3H7mltZqocIOFJHRcciJMcnIqRlQ9EzIUBBg+FBinpgA3HwCAPR9yABGmGAAFshwzchgOPIQOpNWzjgMCGRKYJPIALziIFKyCsrB6lq6+sosAOSd12w96yCbw7v7h3IQiio2uOd6BhYAEYAEwAZgeDCAA)

**Comment by AnyhowStep** [comment]
Your `foo` has an explicit type of `T extends FactoryFunction<infer U> ? U : never`.
So, even if it resolves to `any`, it still isn't "implicit".

**Comment by AnyhowStep** [comment]
That aside, your type looks very suspicious. You're not actually using `T` in a way that TS can use for inference. So, it's just defaulting to the constraint type.

**Comment by RyanCavanaugh** [comment]
This is the intended behavior per @AnyhowStep 's comments. If inference ends up with no candidates, it checks the call as if you had passed the constraints as type arguments.

I don't know what behavior you're trying to accomplish, but this isn't the right way to do it.
